### PR TITLE
fix(rest): support rewinds in libcurl

### LINKED ISSUE #11167: storage: Send failed since rewinding of the data stream failed
https://console.cloud.google.com/cloud-build/builds;region=us-east1/295302d8-cc54-4579-9b4a-dad5adc53d72?project=936212892354

UNKNOWN: Permanent error CreateResumableUpload: PerformWork() - CURL error [65]=Send failed since rewinding of the data stream failed

```
Step #3: google/cloud/storage/tests/object_parallel_upload_integration_test.cc:56: Failure
Step #3: Value of: object_metadata
Step #3: Expected: code is equal to OK and message is anything
Step #3:   Actual: 848-byte object <C0-BC 48-02 00-00 00-00 00-5E EC-05 E4-59 9D-8B 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 ... 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00>, whose status is UNKNOWN: Permanent error CreateResumableUpload: PerformWork() - CURL error [65]=Send failed since rewinding of the data stream failed, with a code that isn't equal to OK, but a message that is anything
```
```
Step #3: 2023-04-03T23:04:08.350726585Z [INFO] <139623562084480> CreateResumableUpload() << ResumableUploadRequest={bucket_name=gcs-grpc-team-cloud-cpp-testing-bucket, object_name=ssyicrkwiwafldgr.upload_shard_22, disable-md5-hash=1, object-metadata=ObjectMetadata={name=, acl=[], bucket=, cache_control=, component_count=0, content_disposition=, content_encoding=, content_language=, content_type=application/binary, crc32c=, etag=, event_based_hold=false, generation=0, id=, kind=, kms_key_name=, md5_hash=, media_link=, , metageneration=0, name=, retention_expiration_time=1970-01-01T00:00:00Z, self_link=, size=0, storage_class=, temporary_hold=false, time_created=0, time_deleted=0, time_storage_class_updated=0, updated=0}, auto-finalize=enabled} (google/cloud/storage/internal/logging_client.cc:44)
Step #3: 2023-04-03T23:04:08.400304974Z [INFO] <139623562084480> CreateResumableUpload() >> payload={CreateResumableUploadResponse={upload_id=https://storage.googleapis.com/upload/storage/v1/b/gcs-grpc-team-cloud-cpp-testing-bucket/o?uploadType=resumable&upload_id=ADPycdu9lAqngZNtuJNoehDdXAMGaf4i1o5YEaYx_rm1d73meXUC5UVab75l7HM-ku3joCxwoVpAkyVJzQ9Y1KcXo8EZCfyC-Yig}} (google/cloud/storage/internal/logging_client.cc:47)
Step #3: 2023-04-03T23:04:08.402260151Z [INFO] <139623562084480> CreateResumableUpload() << ResumableUploadRequest={bucket_name=gcs-grpc-team-cloud-cpp-testing-bucket, object_name=ssyicrkwiwafldgr.upload_shard_23, disable-md5-hash=1, object-metadata=ObjectMetadata={name=, acl=[], bucket=, cache_control=, component_count=0, content_disposition=, content_encoding=, content_language=, content_type=application/binary, crc32c=, etag=, event_based_hold=false, generation=0, id=, kind=, kms_key_name=, md5_hash=, media_link=, , metageneration=0, name=, retention_expiration_time=1970-01-01T00:00:00Z, self_link=, size=0, storage_class=, temporary_hold=false, time_created=0, time_deleted=0, time_storage_class_updated=0, updated=0}, auto-finalize=enabled} (google/cloud/storage/internal/logging_client.cc:44)
Step #3: 2023-04-03T23:04:08.415173089Z [INFO] <139623562084480> CreateResumableUpload() >> status={UNKNOWN: PerformWork() - CURL error [65]=Send failed since rewinding of the data stream failed} (google/cloud/storage/internal/logging_client.cc:49)
```

### PR DESCRIPTION
**Pull request by coryan**
This fixes problems I introduced in the first PR.

See #11703. Really fixes #11167

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11709)

### CODE REVIEW

**Review by scotthart — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11709)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`google/cloud/internal/curl_writev.h` line 57 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11709#-NW8gLe7EmO2Lqc-Wsx6:-NW8gLe7EmO2Lqc-Wsx7:b-2flqtd) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b611c08900cba72d9770bb45ca8055175b2938d/google/cloud/internal/curl_writev.h#L57)):*
> ```objc
>     }
>     return avail - dst.size();
>   }
> ```

nit: Should these method implementations be moved to the new .cc file or is there an advantage to leaving them here in the .h?



_Code quote:_
```objc
  explicit WriteVector(std::vector<absl::Span<char const>> v)
      : original_(std::move(v)), writev_(original_.begin(), original_.end()) {}

  std::size_t size() const {
    std::size_t size = 0;
    for (auto const& s : writev_) {
      size += s.size();
    }
    return size;
  }

  std::size_t MoveTo(absl::Span<char> dst) {
    auto const avail = dst.size();
    while (!writev_.empty()) {
      auto& src = writev_.front();
      if (src.size() > dst.size()) {
        std::copy(src.begin(), src.begin() + dst.size(), dst.begin());
        src.remove_prefix(dst.size());
        dst.remove_prefix(dst.size());
        break;
      }
      std::copy(src.begin(), src.end(), dst.begin());
      dst.remove_prefix(src.size());
      writev_.pop_front();
    }
    return avail - dst.size();
  }
```

File: google/cloud/internal/curl_writev.cc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include "google/cloud/internal/curl_writev.h"
+#include <cstdio>
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+bool WriteVector::Seek(std::size_t offset, int origin) {
+  // libcurl claims to only req
```
Comment (devbww) [inline google/cloud/internal/curl_writev.cc:24]: Incomplete comment?

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/curl_writev.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/curl_writev.cc]: Fixed

**Review by coryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11709)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @devbww and @scotthart)

___
*[`google/cloud/internal/curl_writev.h` line 57 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11709#-NW8gLe7EmO2Lqc-Wsx6:-NW9C6ht79QVuobuJch6:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b611c08900cba72d9770bb45ca8055175b2938d/google/cloud/internal/curl_writev.h#L57)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

nit: Should these method implementations be moved to the new .cc file or is there an advantage to leaving them here in the .h?
</blockquote></details>

Done.

**Pull request merged**